pkg/restore/cdclog/puller.go

```diff
@@ -0,0 +1,177 @@
+// Copyright 2020 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cdclog
+
+import (
+	"context"
+
+	"github.com/pingcap/errors"
+	"github.com/pingcap/log"
+
+	"github.com/pingcap/br/pkg/storage"
+)
+
+// EventPuller pulls next event in ts order.
+type EventPuller struct {
+	ddlDecoder            *JSONEventBatchMixedDecoder
+	rowChangedDecoder     *JSONEventBatchMixedDecoder
+	currentDDLItem        *SortItem
+	currentRowChangedItem *SortItem
+
+	schema string
+	table  string
+
+	storage         storage.ExternalStorage
+	ddlFiles        []string
+	rowChangedFiles []string
+
+	ddlFileIndex        int
+	rowChangedFileIndex int
+}
+
+// NewEventPuller create eventPuller by given log files, we assume files come in ts order.
+func NewEventPuller(
+	ctx context.Context,
+	schema string,
+	table string,
+	ddlFiles []string,
+	rowChangedFiles []string,
+	storage storage.ExternalStorage) (*EventPuller, error) {
+	var (
+		ddlDecoder        *JSONEventBatchMixedDecoder
+		ddlFileIndex      int
+		rowChangedDecoder *JSONEventBatchMixedDecoder
+		rowFileIndex      int
+	)
+	if len(ddlFiles) == 0 {
+		log.Info("There is no ddl file to restore")
+	} else {
+		data, err := storage.Read(ctx, ddlFiles[0])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		ddlFileIndex++
+		ddlDecoder, err = NewJSONEventBatchDecoder(data)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	if len(rowChangedFiles) == 0 {
+		log.Info("There is no row changed file to restore")
+	} else {
+		data, err := storage.Read(ctx, rowChangedFiles[0])
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+		rowFileIndex++
+		rowChangedDecoder, err = NewJSONEventBatchDecoder(data)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+
+	return &EventPuller{
+		schema: schema,
+		table:  table,
+
+		ddlDecoder:        ddlDecoder,
+		rowChangedDecoder: rowChangedDecoder,
+
+		ddlFiles:            ddlFiles,
+		rowChangedFiles:     rowChangedFiles,
+		ddlFileIndex:        ddlFileIndex,
+		rowChangedFileIndex: rowFileIndex,
+
+		storage: storage,
+	}, nil
+}
+
+// PullOneEvent pulls one event in ts order.
+// The Next event which can be DDL item or Row changed Item depends on next commit ts.
+func (e *EventPuller) PullOneEvent(ctx context.Context) (*SortItem, error) {
```

Comment:
This function actually implement a `mergingIterator` with heap length 2, the big block code for row/dll agains look quite the same